### PR TITLE
Cys bridges

### PR DESCRIPTION
**Pull request by jbarnoud**
Implement disulfide bridges. The main issue prior to this PR is that the link does not have a way to describe two atoms from two distinct residues having both the same resname and atomname, but an unknown distance in resid. The system based on numerical `order` means we have to know before hand how may residue will separate the two residues incvolved in the cys-bridge.

This PR extend the 'order' attribute with series of '<', '>', and '*' in addition to the numerical values. The details are described in the docstring of `vermouth.processors.do_links._match_order`.

The change is reflected in the ff parser and in the force field files. A series of test are added to valisate the order reading and comparison.

The feature still requires the implementation of some graph manipulation, and requires additions to the CLI. For now, the edge for a disulfide bridge is set by `MakeBonds`; it comes more or less to the "auto" option in Martinize 1. The default in martinize1 is however not to have bridges, which means we need a processor able to remove the edges corresponding to these bridges, but also to create edges based on a user defined distance. Finally, we need to reimplement the selecion method to be able to create a specific edge. That last processor, if written correctly, can also create edges in other context, like between a ligand and a protein.

### CODE REVIEW

File: vermouth/processors/tune_cystein_bridges.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        for key in base_molecule:
+            correspondance[(base_index, key)] = (new_index, key)
+        offset = max(base_molecule.nodes)
+
+        for other_index in component[1:]:
+            other_molecule = molecules[other_index]
+            mol_correspondance = base_molecule.merge_molecule(other_molecule)
+            for before, after in mol_correspondance.items():
+                correspondance[(other_index, before)] = (new_index, after)
+            offset = max(base_molecule.nodes)
```
Comment (pckroon) [inline vermouth/processors/tune_cystein_bridges.py]: offset is not used, also above

File: vermouth/processors/tune_cystein_bridges.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+    Yields
+    ------
+    tuple
+        Pairs of node closer than the threshold in the format described above.
+
+    Raises
+    ------
+    KeyError
+        Raised if a node from the selection does not have a position.
```
Comment (pckroon) [inline vermouth/processors/tune_cystein_bridges.py]: Same as before

File: vermouth/tests/test_links.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    assert match == answer
+
+
+@pytest.mark.parametrize("order", (
+    1.2, -3.9, None, True, False, {'a': 3},
+    '><', '#>>', '#$H', '', [], np.nan, 
+))
+def test_order_errors(order):
+    with pytest.raises(ValueError):
+        _, _ = do_links._interpret_order(order)
```
Comment (pckroon) [inline vermouth/tests/test_links.py]: Don't assign to variable

File: vermouth/processors/do_links.py [hunk 4]
```diff
@@ -72,6 +74,155 @@ def _pattern_match(molecule, atoms, raw_match):
...
+            # Column >, >>, <, and <<
+            return False
+    elif order_types[0] == '*':  # Rows *, and **
+        if order_types[1] == 'number' and orders[1] == 0 and resid1 == resid2:
+            # Column 0
+            return False
+        elif order_types[1] == '*' and ((orders[0] == orders[1]) != (resid1 == resid2)):
+            # Columns *, and **
+            return False
+
```
[the remark this answers is not in the document]
Comment (pckroon) [reply vermouth/processors/do_links.py]: Ha! No, it's there. #brainfart

File: vermouth/processors/tune_cystein_bridges.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    # Do the merging
+    new_molecules = []
+    correspondance = {}
+    for new_index, component in enumerate(components):
+        # Connected components are yielded as sets. We need to be able to
+        # iterate over them in numerical order.
+        component = sorted(component)
+
+        # The first molecule in the component is the base for the merge.
+        base_index = component[0]
```
[the remark this answers is not in the document]
Comment (jbarnoud) [reply vermouth/processors/tune_cystein_bridges.py]: why?

File: vermouth/processors/do_links.py [hunk 6]
```diff
@@ -72,6 +74,155 @@ def _pattern_match(molecule, atoms, raw_match):
...
+    order_types = []
+    for i, order in enumerate((order1, order2), start=1):
+        order_type, order_value = _interpret_order(order)
+        order_types.append(order_type)
+        orders.append(order_value)
+
+    if order_types[0] == 'number':  # Rows n and 0 in the comparison matrix
+        if order_types[1] == 'number':
+            # Columns n, and 0
+            if (orders[1] - orders[0]) != (resid2 - resid1):
```
[the remark this answers is not in the document]
Comment (jbarnoud) [reply vermouth/processors/do_links.py]: The orders and the resids must be in the same order. Using the absolute value looses that information.

File: vermouth/processors/do_links.py [hunk 7]
```diff
@@ -72,6 +74,155 @@ def _pattern_match(molecule, atoms, raw_match):
...
+            order_value = signs[first_character] * len(order)
+        elif first_character == '*':
+            # This could be an 'else', but it would hide bugs if the code
+            # above changes.
+            order_type = '*'
+            order_value = len(order)
+    return order_type, order_value
+
+
+def _match_order(order1, resid1, order2, resid2):
```
[the remark this answers is not in the document]
Comment (jbarnoud) [reply vermouth/processors/do_links.py]: Where would it be used? What is the benefit of exposing the function in the API?

File: vermouth/processors/do_links.py [hunk 8]
```diff
@@ -72,6 +74,155 @@ def _pattern_match(molecule, atoms, raw_match):
...
+    | << | < | <  | < | =  | ! | < | ! | !  |
+    +----+---+----+---+----+---+---+---+----+
+    | n  | ! | !  | ! | !  | ? | ? | ! | !  |
+    +----+---+----+---+----+---+---+---+----+
+    | 0  | < | <  | > | >  | ? | = | / | /  |
+    +----+---+----+---+----+---+---+---+----+
+    | *  | ! | !  | ! | !  | ! | / | = | /  |
+    +----+---+----+---+----+---+---+---+----+
+    | ** | ! | !  | ! | !  | ! | / | / | =  |
+    +----+---+----+---+----+---+---+---+----+
```
[the remark this answers is not in the document]
Comment (jbarnoud) [reply vermouth/processors/do_links.py]: The `=` are valid sphinx. I am not sure the `||` is.

File: vermouth/ffinput.py [hunk 9]
```diff
@@ -284,53 +286,172 @@ def _treat_block_interaction_atoms(atoms, context, section):
...
+                          'of ">", "<", or "*".'
+                          .format(order))
+    return prefix_from_attributes, order_from_attributes
+
+
+def _get_order_and_prefix_from_prefix(prefix):
+    prefix_from_prefix = None
+    order_from_prefix = 0
+    if prefix:
+        # It is already validated.
```
[the remark this answers is not in the document]
Comment (jbarnoud) [reply vermouth/ffinput.py]: I do not like having twice the same line with the returns.

[the remark this answers is not in the document]
Comment (pckroon) [reply vermouth/ffinput.py]:
```python 
def _get_order_and_prefix_from_prefix(prefix):
    if not prefix:
        return None, 0
    # It is already validated.
    if prefix[0] == '+':
        order_from_prefix = len(prefix)
    elif prefix[0] == '-':
        order_from_prefix = -len(prefix)
    else:
        order_from_prefix = prefix

    return prefix, order_from_prefix

File: vermouth/processors/do_links.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pckroon) [reply vermouth/processors/do_links.py]: That's the point of having an API. People can solve problems we don't forsee with tools we provide,

File: vermouth/processors/tune_cystein_bridges.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pckroon) [reply vermouth/processors/tune_cystein_bridges.py]: You just want to "lowest" value, so you shouldn't sort things. Maybe `base_molecule = min(molecules, key=len)` or something. I need to stare for a little too long to figure it out exactly for now.

[the remark this answers is not in the document]
Comment (jbarnoud) [reply vermouth/processors/tune_cystein_bridges.py]: I do not only want the lowest; latter I iterate over them in order.

**Comment by jbarnoud** [comment]
There are a couple of things missing, but they can wait after the feature freeze:

* pruning the bridges does not separate the molecules, but instead leave one disconnected molecule;
* martinize 1 allows to specify individual cys bridges, martinize 2 cannot yet.

I rebased against master.

**Comment by pckroon** [comment]
Could you make issues for the two missing features? I'll merge this in the meantime. Nice work.

**Pull request merged**